daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/scenario/ScenarioLedger.scala

```diff
@@ -375,141 +374,172 @@ object ScenarioLedger {
+    * @param ledgerData ledger recording all historical transaction that have been processed
+    * @return updated ledger with new transaction information
+    */
   private def processTransaction(
       trId: TransactionId,
       richTr: RichTransaction,
       locationInfo: Map[NodeId, Location],
       ledgerData: LedgerData,
   ): Either[UniqueKeyViolation, LedgerData] = {
 
-    final case class RollbackBeginState(
-        activeContracts: Set[ContractId],
-        activeKeys: Map[GlobalKey, ContractId],
-    )
+    def addNewLedgerNodes(historicalLedgerData: LedgerData): LedgerData =
+      richTr.transaction.transaction.reachableNodeIds.foldLeft(historicalLedgerData) {
+        case (ledgerData, nodeId) =>
+          val eventId = EventId(trId.id, nodeId)
+
+          richTr.transaction.nodes.get(nodeId) match {
+            case None =>
+              crash(s"processTransaction: non-existent node '$eventId'.")
+
+            case Some(node) =>
+              val newLedgerNodeInfo = LedgerNodeInfo(
+                node = node,
+                optLocation = locationInfo.get(nodeId),
+                transaction = trId,
+                effectiveAt = richTr.effectiveAt,
+                // Following fields will be updated by additional calls to node processing code
+                disclosures = Map.empty,
+                referencedBy = Set.empty,
+                consumedBy = None,
+                rolledbackBy = None,
+                parent = None,
+              )
 
-    final case class ProcessingNode(
-        // The id of the direct parent or None.
-        mbParentId: Option[NodeId],
-        // The id of the nearest rollback ancestor. If this is
-        // a rollback node itself, it points to itself.
-        mbRollbackAncestorId: Option[NodeId],
-        children: List[NodeId],
-        // For rollback nodes, we store the previous state here and restore it.
-        // For exercise nodes, we don’t need to restore anything.
-        prevState: Option[RollbackBeginState],
-    )
+              ledgerData.copy(nodeInfos = ledgerData.nodeInfos + (eventId -> newLedgerNodeInfo))
+          }
+      }
 
-    @tailrec
-    def processNodes(
-        cache0: LedgerData,
-        enps: List[ProcessingNode],
-    ): LedgerData = enps match {
-      case Nil => cache0
-      case ProcessingNode(_, _, Nil, optPrevState) :: restENPs => {
-        val cache1 = optPrevState.fold(cache0) { case prevState =>
-          cache0.copy(
-            activeContracts = prevState.activeContracts,
-            activeKeys = prevState.activeKeys,
-          )
-        }
-        processNodes(cache1, restENPs)
+    def createdUpdates(historicalLedgerData: LedgerData): LedgerData =
+      richTr.transaction.transaction.reachableNodeIds.foldLeft(historicalLedgerData) {
+        case (ledgerData, nodeId) =>
+          val eventId = EventId(trId.id, nodeId)
+
+          // As `addNewLedgerNodes` has already ran, the following node lookup can not fail
```

Comment:
As the update functions are all nested within `processTransaction`, its maybe not worthwhile refining the LedgerData type to capture that this check has been performed?

Reply:
i don’t think it’s worth the complexity tbh